flowjax/train/variational_fit.py

```diff
@@ -0,0 +1,89 @@
+from typing import Optional, Callable
+
+from jax import random
+import optax
+import equinox as eqx
+from tqdm import tqdm
+
+from flowjax.utils import Array
+from flowjax.distributions import Distribution
+
+# Variational targets accept an array of proposed samples and return an array of (usually)
+# the unormalised log posterior probability of the samples.
+VariationalTarget = Callable[[Array], Array]
+
+# Variational losses are functions that take a distribution, a target 
+# callable and a random key, and returns a scalar loss.
+VariationalLoss = Callable[
+    [Distribution, VariationalTarget, random.KeyArray], 
+    float
+]
+
+@eqx.filter_jit
+def elbo_loss(dist: Distribution, target: VariationalTarget, key: random.KeyArray, elbo_samples: int = 500):
+    samples = dist.sample(key, sample_shape=(elbo_samples,))
+    approx_density = dist.log_prob(samples).reshape(-1)
+    target_density = target(samples).reshape(-1)
+    losses = approx_density - target_density
+    return losses.mean()
+
+def variational_fit(
+    key: random.KeyArray,
+    dist: Distribution,
+    target: VariationalTarget,
+    loss_fcn: VariationalLoss = elbo_loss,
```

Comment:
In other places in the code I use `fn` to abbreviate function, so would be nice to keep it consistent here